### PR TITLE
Prometheus metrics

### PR DESCRIPTION
**Pull request by kostekIV**
We are using your proxy for quite some time now and it works great for us, thank you for this :)
We needed metrics for our internal reporting & stuff so quite a while ago I have added them to our fork of subway. I also saw you have them marked as a TODO in the readme. So this pr is a work I have already done to introduce prometheus metrics to subway. If you think this would be nice to have here you go, of course Im happy to go through rounds of review :) 

This pr contains 2 major changes. 
1. jsonrpsee version bump 
2. Adding prometheus metrics for methods, ws/http connections count and cache + startup of the prometheus exporter

To see changes related only to `jsonrpsee` bump look at the commit `update jsonrpsee` and simiraly to see chanegs related only to prometheus inclusion take a look at the `Add prometheus metrics`

### Why jsonrpsee bump?
While it was not strictly need, without it the metrics related to the ws connection would be nearly impossible to write (and for sure it would be quite painfull to do).
Also since I did not have an access to your fork of the jsonrpsee this was the quickest thing I could do to make those changes.

### What kind of metrics there are?
So there is ws/http counter metrics for open/closed sessions. Also metrics for the methods that are based on the ones that are used in the substrate - same names etc. Additionally metrics for cache misses per method.

Cc @xlc

### CODE REVIEW

**Comment by xlc** [comment]
Thanks for the PR! Will take a review next week.

**Comment by xlc** [comment]
And it will be great if you can get the CI pass meanwhile. I have updated the toolchain version on master.

File: src/extensions/prometheus/mod.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    pub fn new(config: PrometheusConfig) -> Self {
+        let labels = config
+            .chain_label
+            .clone()
+            .map(|l| iter::once(("chain".to_string(), l.clone())).collect());
+        let prefix = match config.prefix {
+            Some(p) if p.is_empty() => None,
+            p => p,
+        };
+        let registry = Registry::new_custom(prefix, labels).expect("Can't happen");
```
Comment (xlc) [inline src/extensions/prometheus/mod.rs:63]: why it can't fail?

File: src/extensions/prometheus/mod.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        &self.registry
+    }
+
+    pub fn rpc_metrics(&self) -> RpcMetrics {
+        self.rpc_metrics.clone()
+    }
+}
+
+fn start_prometheus_exporter(registry: Registry, port: u16, listen_address: String) -> JoinHandle<()> {
+    let address = listen_address.parse().unwrap();
```
Comment (xlc) [inline src/extensions/prometheus/mod.rs:84]:
```suggestion
    let address = listen_address.parse().expect("Invalid prometheus listen address");
```

File: src/extensions/prometheus/mod.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    let addr = SocketAddr::new(address, port);
+
+    tokio::spawn(async move {
+        init_prometheus(addr, registry).await.unwrap();
```
Comment (xlc) [inline src/extensions/prometheus/mod.rs:88]:
```suggestion
        init_prometheus(addr, registry).await.expect("Init prometeus failed");
```

File: src/extensions/server/prometheus.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+pub enum Protocol {
+    Ws,
+    Http,
+}
+
+impl Display for Protocol {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let str = match self {
+            Self::Ws => "ws".to_string(),
+            Self::Http => "http".to_string(),
```
Comment (xlc) [inline src/extensions/server/prometheus.rs:19]:
```suggestion
            Self::Ws => "ws",
            Self::Http => "http",
```

File: src/extensions/prometheus/mod.rs (hunk 1 above)
Comment (kostekIV) [reply src/extensions/prometheus/mod.rs:63]: It only can fail if the prefix is equal to `Some(p) where p.is_mepty()` and we take care of that in the match above

Comment (xlc) [reply src/extensions/prometheus/mod.rs:63]: can you explain it in the `expect`? otherwise the next reader is likely going to have the same question

Comment (kostekIV) [reply src/extensions/prometheus/mod.rs]: Sure, I think now it should be better

**Review by xlc — APPROVED** [review]

**Pull request merged**